internal/thea.go

```diff
@@ -3,323 +3,196 @@ package internal
-// go-routine, ensuring that the Thea service waitgroup is updated correctly
-func (thea *theaImpl) spawnAsyncService(service func()) {
-	thea.shutdownWaitGroup.Add(1)
 
-	go func() {
-		defer thea.shutdownWaitGroup.Done()
-		service()
-	}()
-}
-
-// synchroniseQueue will first discover all items inside the import directory,
-// and then will injest any that do not already exist in the queue. Any items
-// in the queue that no longer exist in the discovered items will also be cancelled
-func (thea *theaImpl) synchroniseQueue() error {
-	// Find new items
-	thea.queueMgr.Reload()
-	presentItems, err := thea.discoverItems()
-	if err != nil {
+	log.Emit(logger.NEW, "Connecting to database...\n")
+	db := database.New()
+	if store, err := NewStoreOrchestrator(db); err == nil {
+		thea.storeOrchestrator = store
+	} else {
 		return err
 	}
 
-	for path, info := range presentItems {
-		thea.queueMgr.Push(queue.NewQueueItem(info, path, thea))
+	if err := db.Connect(thea.config.Database); err != nil {
+		return err
+	}
+	searcher := tmdb.NewSearcher(tmdb.Config{ApiKey: thea.config.OmdbKey})
+	scraper := media.NewScraper(media.ScraperConfig{FfprobeBinPath: thea.config.Format.FfprobeBinaryPath})
+	if serv, err := ingest.New(thea.config.IngestService, searcher, scraper, thea.storeOrchestrator, thea.eventBus); err == nil {
+		thea.ingestService = serv
+	} else {
+		panic(fmt.Sprintf("failed to construct ingestion service due to error: %s", err.Error()))
 	}
 
-	thea.queueMgr.Filter(func(queue queue.QueueManager, key int, item *queue.Item) bool {
-		if _, ok := presentItems[item.Path]; !ok {
-			thea.CancelItem(item.ItemID)
-			return false
-		}
-
-		return true
-	})
-
-	thea.queueMgr.ForEach(func(q queue.QueueManager, idx int, item *queue.Item) bool {
-		if item.Stage != queue.Import {
-			return false
-		}
-
-		info, err := os.Stat(item.Path)
-		if err != nil {
-			log.Emit(logger.WARNING, "Failed to get file info for %v during import stage: %v\n", item.Path, err.Error())
-			return false
-		}
+	if serv, err := transcode.New(thea.config.Format, thea.eventBus, thea.storeOrchestrator); err == nil {
+		thea.transcodeService = serv
+	} else {
+		panic(fmt.Sprintf("failed to construct transcode service due to error: %s", err.Error()))
+	}
 
-		if time.Since(info.ModTime()) > time.Minute*2 {
-			log.Emit(logger.INFO, "Advancing item %s from Import hold as it's exceeded modtime threshold\n", item)
-			thea.AdvanceItem(item)
-		}
+	thea.restGateway = api.NewRestGateway(&thea.config.RestConfig, thea.ingestService, thea.transcodeService, thea.storeOrchestrator)
+	thea.activityManager = newActivityManager(thea.restGateway, thea.eventBus)
 
-		return false
-	})
+	wg := &sync.WaitGroup{}
+	thea.spawnAsyncService(ctx, wg, thea.ingestService, "ingest-service", crashHandler)
+	thea.spawnAsyncService(ctx, wg, thea.transcodeService, "transcode-service", crashHandler)
+	thea.spawnAsyncService(ctx, wg, thea.restGateway, "rest-gateway", crashHandler)
```

Comment:
I don't know if this is a me thing or a Go thing, but I would usually lift the `go` out of this function. So it looks more like this:

```suggestion
	wg := &sync.WaitGroup{}
    wg.Add(3)
	go thea.spawnAsyncService(ctx, wg, thea.ingestService, "ingest-service", crashHandler)
	go thea.spawnAsyncService(ctx, wg, thea.transcodeService, "transcode-service", crashHandler)
	go thea.spawnAsyncService(ctx, wg, thea.restGateway, "rest-gateway", crashHandler)
```

Reply:
Hm yeah I can definitely see the benefit, which is that it makes it very clear to the reader that the service is being spawned in it's own goroutine (which would mean I could remove the `Async` from the name too). Makes the spawn function much cleaner too!